### PR TITLE
Fix language in documentation referring to txn retries

### LINKED ISSUE #1117: storage: push txn queues for high contention scenarios
Important updates to how we handle transaction priority in high-contention scenarios. Need to update `transactions.md` and other transaction-related docs.

PR: https://github.com/cockroachdb/cockroach/pull/13501

cc: @spencerkimball

### PR DESCRIPTION
**Pull request by spencerkimball**
Fixes #1117

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/1139)

### CODE REVIEW

**Comment by jseldess** [comment]
Thanks for making these updates, @spencerkimball! Almost everything looks good, but I'm still confused about some language around increasing transaction priority in `transactions.md`. Can you help clarify for me? 

---

Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[transactions.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeQ7X8mrbr2ygHgGzN0:-KeQ7X8mrbr2ygHgGzN1:bz7dgtl) ([raw file](https://github.com/cockroachdb/docs/blob/fa9a7a1af8ec73ce9aaaee67ce765bc8a657f3e0/transactions.md#L9)):*
> ```Markdown
> CockroachDB supports bundling multiple SQL statements into a single all-or-nothing transaction. Each transaction guarantees [ACID semantics](https://en.wikipedia.org/wiki/ACID) spanning arbitrary tables and rows, even when data is distributed. If a transaction succeeds, all mutations are applied together with virtual simultaneity. If any part of a transaction fails, the entire transaction is aborted, and the database is left unchanged. CockroachDB guarantees that while a transaction is pending, it is isolated from other concurrent transactions with either snapshot or serializable [isolation levels](#isolation-levels).
> 
> {{site.data.alerts.callout_info}}For a detailed discussion of CockroachDB transaction semantics, see <a href="https://www.cockroachlabs.com/blog/how-cockroachdb-distributes-atomic-transactions/">How CockroachDB Does Distributed Atomic Transactions</a> and <a href="https://www.cockroachlabs.com/blog/serializable-lockless-distributed-isolation-cockroachdb/">Serializable, Lockless, Distributed: Isolation in CockroachDB</a>.{{site.data.alerts.end}} Note that the explanation of the transaction model described in this blog post is slightly out of date. See the [Transaction Retries](#transaction-retries) section for more details.
> ```

nit: Put this new note on a new line, unless you want it to be part of the alert; in that case, it should be placed before `{{site.data.alerts.end}}`.

---

*[transactions.md, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeQB4R6ndRFoeOwnHMD:-KeQB4R8548WjpsJ13TB:bmed9md) ([raw file](https://github.com/cockroachdb/docs/blob/fa9a7a1af8ec73ce9aaaee67ce765bc8a657f3e0/transactions.md#L197)):*
> ```Markdown
> ~~~
> 
> {{site.data.alerts.callout_info}}When two transactions contend for the same resources indirectly, they may create a dependency cycle leading to a deadlock situation, where both transactions are waiting on each other to finish. In these cases, CockroachDB allows the transaction with higher priority to abort the other, which must retry. On retry, the transaction inherits the higher priority. This means that each retry makes a transaction more likely to succeed in the event it again experiences deadlock.{{site.data.alerts.end}}
> ```

I think I'm missing something. Here and in a few other places, we say that the transaction with the higher priority aborts the other transaction. But isn't it now the case that the transaction accessing the resource first locks it, which causes the contending transaction to abort and retry? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-:-KeQBjWFbrZdTm71x06Y:b-jff02f)*

**Comment by spencerkimball** [comment]
Let me know if there's still confusion in the document itself, of if you were just confused about the abort behavior with the new model.

---

Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[transactions.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeQ7X8mrbr2ygHgGzN0:-KeVw1GoEtkb_s2gKlRU:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/fa9a7a1af8ec73ce9aaaee67ce765bc8a657f3e0/transactions.md#L9)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

nit: Put this new note on a new line, unless you want it to be part of the alert; in that case, it should be placed before `{{site.data.alerts.end}}`.
</blockquote></details>

Done.

---

*[transactions.md, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeQB4R6ndRFoeOwnHMD:-KeVwI4sCRDkFGC_bK3v:b-xyaylu) ([raw file](https://github.com/cockroachdb/docs/blob/fa9a7a1af8ec73ce9aaaee67ce765bc8a657f3e0/transactions.md#L197)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

I think I'm missing something. Here and in a few other places, we say that the transaction with the higher priority aborts the other transaction. But isn't it now the case that the transaction accessing the resource first locks it, which causes the contending transaction to abort and retry? 
</blockquote></details>

We no longer abort and retry just because another transaction has locked a resource. We instead wait for the resource to be unlocked and then proceed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-:-KeVwROiIHWpM9PzNvs_:b61k4tp)*

**Comment by jseldess** [comment]
Thanks for the clarification, Spencer. I think there's still confusion in the document itself. Specifically, there's still a lot of emphasis on retries and priorities, which, for me at least, distorts the picture a bit. Here are some ideas:

- It would help to expand the intro to briefly mention conflicts, that we lock and have the contending transaction wait until the resource is free unless there's deadlock, in which case we abort, increase priority, and retry. 
- Perhaps we need a new section, before the content about retries and priorities, explaining how we handle contention normally. This would give context to the retry and priorities sections.
- We may also need to reframe the retry and priority content itself a little to make sure it's clear that we retry and increase priority only when there's deadlock. 

Thoughts? If you prefer, I could work on this in a follow-up PR.

---

Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-:-KeZB5iMT6p8X29jogTs:b-lz0dfi)*

**Comment by bdarnell** [comment]
Review status: 0 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[begin-transaction.md, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZIihb8VyE0oIiybcu:-KeZIihb8VyE0oIiybcv:b-80oewm) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/begin-transaction.md#L9)):*
> ```Markdown
> The `BEGIN` [statement](sql-statements.html) initiates a [transaction](transactions.html), which either successfully executes all of the statements it contains or none at all.
> 
> {{site.data.alerts.callout_danger}}When using transactions, your application should include logic to <a href="transactions.html#transaction-retries">retry transactions that are aborted to break a dependency cycle between concurrent transactions leading to deadlock</a>.{{site.data.alerts.end}}
> ```

You have to retry for read/write contention, not just deadlocks. 

---

*[frequently-asked-questions.md, line 109 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZJ-WRVIXrYtGeBsuV:-KeZJ-WRVIXrYtGeBsuW:b-h7gawt) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/frequently-asked-questions.md#L109)):*
> ```Markdown
> [Transactions](transactions.html) in CockroachDB do not explicitly lock all of their data resources. Instead, using [optimistic concurrency control (OCC)](https://en.wikipedia.org/wiki/Optimistic_concurrency_control), CockroachDB proceeds with transactions under the assumption that there’s no contention until commit time. In cases without contention, this results in higher performance than explicit locking would allow. With contention, transactions may need to be retried.
> 
> In practice, most applications experience low contention. However, with significant contention, OCC may perform poorly. If your application experiences high rates of contention, snapshot isolation (SI) can significantly improve performance by eliminating the possibility of retries (except in cases of deadlock).
> ```

Is deadlock really the only reason SI transactions can be restarted? I think there's still ReadWithinUncertaintyInterval and a few other things (like coordinator restarts)

---

*[transactions.md, line 7 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZJfyyh56ys6RoR0RR:-KeZJfyzrICbsNo4k5CD:bikq6r4) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L7)):*
> ```Markdown
> ---
> 
> CockroachDB supports bundling multiple SQL statements into a single all-or-nothing transaction. Each transaction guarantees [ACID semantics](https://en.wikipedia.org/wiki/ACID) spanning arbitrary tables and rows, even when data is distributed. If a transaction succeeds, all mutations are applied together with virtual simultaneity. If any part of a transaction fails, the entire transaction is aborted, and the database is left unchanged. CockroachDB guarantees that while a transaction is pending, it is isolated from other concurrent transactions with either snapshot or serializable [isolation levels](#isolation-levels).
> ```

"with either snapshot or serializable isolation levels" sounds like there are other isolation levels that don't provide this isolation. I think I'd remove this clause, or reword it to make it more clear that you're providing a definition of "isolation" instead of qualifying "other transactions". 

---

*[transactions.md, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZKpWrx3KvAFqWWLCN:-KeZKpWrx3KvAFqWWLCO:b-hu2gr1) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L68)):*
> ```Markdown
> ## Transaction Retries
> 
> Transactions in CockroachDB lock data resources which are written during their execution. In the event that a pending write from one transaction conflicts with a read or a write of a concurrent transaction, then the concurrent transaction must wait for the earlier transaction to complete before proceeding. CockroachDB implements a distributed deadlock detection algorithm to discover dependency cycles. Deadlocks are resolved by allowing transactions with higher priority to abort their dependencies. Transactions which are aborted to avoid deadlock must be retried.
> ```

s/read or a write/write/

---

*[transactions.md, line 70 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZL5BWFV-Q9GGwzeko:-KeZL5BWFV-Q9GGwzekp:bmae2s2) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L70)):*
> ```Markdown
> Transactions in CockroachDB lock data resources which are written during their execution. In the event that a pending write from one transaction conflicts with a read or a write of a concurrent transaction, then the concurrent transaction must wait for the earlier transaction to complete before proceeding. CockroachDB implements a distributed deadlock detection algorithm to discover dependency cycles. Deadlocks are resolved by allowing transactions with higher priority to abort their dependencies. Transactions which are aborted to avoid deadlock must be retried.
> 
> Transactions executed with the serializable isolation level may also require retries if they experience read/write contention with other concurrent transactions, though this is limited to a single retry. Note that these types of transaction retries do not occur for transactions executing with the snapshot isolation level. Consider using snapshot isolation if your use case has high contention and your clients are retrying frequently. See [Isolation Levels](#isolation-levels) for more details.
> ```

How is this limited to a single retry?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1139)*

**Comment by spencerkimball** [comment]
Sounds like you're going to take over this PR. I'll address Ben's comments in Reviewable here.

---

Review status: 0 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[begin-transaction.md, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZIihb8VyE0oIiybcu:-Ke_ogcZWS1mdlvL4ue5:b-uwsif2) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/begin-transaction.md#L9)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

You have to retry for read/write contention, not just deadlocks. 
</blockquote></details>

Yes, let's make this less specific by not mentioning deadlocks.

---

*[frequently-asked-questions.md, line 109 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZJ-WRVIXrYtGeBsuV:-Ke_ozNQDc4_Y2H3jSLE:bx2aieo) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/frequently-asked-questions.md#L109)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is deadlock really the only reason SI transactions can be restarted? I think there's still ReadWithinUncertaintyInterval and a few other things (like coordinator restarts)
</blockquote></details>

You're right. Both of those will cause restarts.

---

*[transactions.md, line 7 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZJfyyh56ys6RoR0RR:-Ke_qUbrEt0pLwvexBc9:b-tqusiu) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L7)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

"with either snapshot or serializable isolation levels" sounds like there are other isolation levels that don't provide this isolation. I think I'd remove this clause, or reword it to make it more clear that you're providing a definition of "isolation" instead of qualifying "other transactions". 
</blockquote></details>

Might as well remove that extra verbiage.

---

*[transactions.md, line 70 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZL5BWFV-Q9GGwzeko:-Ke_qXr0kJEQ6OslYEpS:b-emp2tj) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L70)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

How is this limited to a single retry?
</blockquote></details>

It's not when you include uncertainty restarts or txn coordinators dying. I originally specified that you could still have retries if on the retry we write different keys, but dropped that detail because it felt unnecessarily pedantic. Because there are other ways things should restart, we should just drop the "single retry" detail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-:-Ke_ocXsHnl0aPdJsyvw:b-jk8lsx)*

**Comment by jseldess** [comment]
Review status: 0 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[begin-transaction.md, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZIihb8VyE0oIiybcu:-KhV7xr1qYgDyor2Xuoi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/begin-transaction.md#L9)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Yes, let's make this less specific by not mentioning deadlocks.
</blockquote></details>

Done.

---

*[frequently-asked-questions.md, line 109 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZJ-WRVIXrYtGeBsuV:-KhV7P2tc4GrlJnC_ZV8:b-qp4jr1) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/frequently-asked-questions.md#L109)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

You're right. Both of those will cause restarts.
</blockquote></details>

@bdarnell, I don't understand the two addition causes you mention. Can you propose updated language here?

---

*[transactions.md, line 7 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZJfyyh56ys6RoR0RR:-KhV6RzcAGnZVxiIBMWb:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L7)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Might as well remove that extra verbiage.
</blockquote></details>

Done.

---

*[transactions.md, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZKpWrx3KvAFqWWLCN:-KhV6gl8oDaliHggCIap:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L68)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/read or a write/write/
</blockquote></details>

Done.

---

*[transactions.md, line 70 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZL5BWFV-Q9GGwzeko:-KhV6tUKf4XubSwKV-kV:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/transactions.md#L70)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

It's not when you include uncertainty restarts or txn coordinators dying. I originally specified that you could still have retries if on the retry we write different keys, but dropped that detail because it felt unnecessarily pedantic. Because there are other ways things should restart, we should just drop the "single retry" detail.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1139)*

**Comment by jseldess** [comment]
Reviewed 12 of 12 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[frequently-asked-questions.md, line 109 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1139#-KeZJ-WRVIXrYtGeBsuV:-KhYHtMXip1MREou9ApV:bi0yefz) ([raw file](https://github.com/cockroachdb/docs/blob/175291065859b3941471a1b7418d1c2d3dad8c42/frequently-asked-questions.md#L109)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

@bdarnell, I don't understand the two addition causes you mention. Can you propose updated language here?
</blockquote></details>

After chatting with @bdarnell, since it's misleading to use "lock-free", I'm just going to remove this question from the FAQ. We already touch transactions in "Do transactions in CockroachDB guarantee ACID semantics?" and "Does CockroachDB support distributed transactions?". 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1139)*
